File: Darts/src/burlton/dartzee/code/screen/stats/player/StatisticsTabFinishBreakdown.java
package burlton.dartzee.code.screen.stats.player;

import burlton.core.code.obj.HandyArrayList;
import burlton.core.code.obj.HashMapCount;
import burlton.core.code.util.MathsUtil;
import burlton.dartzee.code.object.Dart;
import burlton.dartzee.code.stats.GameWrapper;
import burlton.desktopcore.code.bean.ScrollTable;
import burlton.desktopcore.code.util.ComponentUtil;
import burlton.desktopcore.code.util.TableUtil.DefaultModel;
import org.jfree.chart.ChartFactory;
import org.jfree.chart.ChartPanel;
import org.jfree.chart.JFreeChart;
import org.jfree.chart.plot.PiePlot;
import org.jfree.data.general.DefaultPieDataset;

import javax.swing.*;
import javax.swing.border.TitledBorder;
import javax.swing.event.ListSelectionEvent;
import javax.swing.event.ListSelectionListener;
import java.awt.*;
import java.util.ArrayList;
import java.util.List;

public class StatisticsTabFinishBreakdown extends AbstractStatisticsTab
										  implements ListSelectionListener
{
	private Integer selectedScore = null;
	
	public StatisticsTabFinishBreakdown()
	{
		super();
		
		setLayout(new GridLayout(1, 3, 0, 0));
		
		tableFavouriteDoublesOther.setTableForeground(Color.RED);
		tableFavouriteDoubles.setSelectionMode(ListSelectionModel.SINGLE_SELECTION);
		tableFavouriteDoublesOther.setSelectionMode(ListSelectionModel.SINGLE_SELECTION);
		tablePanel.setBorder(new TitledBorder(UIManager.getBorder("TitledBorder.border"), "Double Finishes", TitledBorder.LEADING, TitledBorder.TOP, null, new Color(0, 0, 0)));
		add(tablePanel);
		tablePanel.setLayout(new GridLayout(2, 1, 0, 0));
		tablePanel.add(tableFavouriteDoubles);
		tablePanel.add(tableFavouriteDoublesOther);
		add(pieChartPanel);
		
		ListSelectionModel model = tableFavouriteDoubles.getSelectionModel();
		model.setSelectionMode(ListSelectionModel.SINGLE_SELECTION);
		model.addListSelectionListener(this);
	}
	
	private final ScrollTable tableFavouriteDoubles = new ScrollTable();
	private final ScrollTable tableFavouriteDoublesOther = new ScrollTable();
	private final JPanel tablePanel = new JPanel();
	private final ChartPanel pieChartPanel = new ChartPanel(null);
	
	@Override
	public void populateStats()
	{
		setTableVisibility();
		
		DefaultPieDataset dataset = buildFavouriteDoublesData(tableFavouriteDoubles, filteredGames);
		if (includeOtherComparison())
		{
			buildFavouriteDoublesData(tableFavouriteDoublesOther, filteredGamesOther);
		}
		
		JFreeChart pieChart = ChartFactory.createPieChart("Finishes", dataset, true, true, false);
		PiePlot plot = (PiePlot)pieChart.getPlot();
		plot.setLabelGenerator(null);
		pieChartPanel.setChart(pieChart);
	}
	
	private DefaultPieDataset buildFavouriteDoublesData(ScrollTable table, HandyArrayList<GameWrapper> filteredGames)
	{
		DefaultModel model = new DefaultModel();
		model.addColumn("Double");
		model.addColumn("Finishes");
		model.addColumn("%");
		
		DefaultPieDataset dataset = populateFavouriteDoubles(model, filteredGames);
		
		table.setModel(model);
		table.sortBy(2, true);
		
		return dataset;
	}
	
	private DefaultPieDataset populateFavouriteDoubles(DefaultModel model, ArrayList<GameWrapper> filteredGames)
	{
		HashMapCount<Integer> hmScoreToCount = new HashMapCount<>();
		for (int i=0; i<filteredGames.size(); i++)
		{
			GameWrapper game = filteredGames.get(i);
			if (!game.isFinished())
			{
				continue;
			}
			
			List<Dart> darts = game.getDartsForFinalRound();
			Dart finalDart = darts.get(darts.size() - 1);
			int score = finalDart.getScore();
			
			hmScoreToCount.incrementCount(score);
		}
		
		ArrayList<Integer> doubles = hmScoreToCount.getKeysAsVector();
		for (int i=0; i<doubles.size(); i++)
		{
			int score = doubles.get(i);
			int count = hmScoreToCount.get(score);
			
			double total = hmScoreToCount.getTotalCount();
			double percent = MathsUtil.getPercentage(count, total);	
			
			//Object[] row = {score, count, percent, color};
			Object[] row = {score, count, percent};
			model.addRow(row);
		}
		
		//Build up the pie set. Unlike the table, we need ALL values
		DefaultPieDataset dataset = new DefaultPieDataset();
		for (int i=1; i<=20; i++)
		{
			int count = hmScoreToCount.getCount(i);
			dataset.setValue(Integer.valueOf(i), count);
		}
		
		dataset.setValue(Integer.valueOf(25), hmScoreToCount.getCount(25));
		return dataset;
	}
	
	private void setTableVisibility()
	{
		if (!includeOtherComparison())
		{
			tablePanel.setLayout(new GridLayout(1, 1, 0, 0));
			tablePanel.remove(tableFavouriteDoublesOther);
		}
		else if (!ComponentUtil.containsComponent(tablePanel, tableFavouriteDoublesOther))
		{
			tablePanel.setLayout(new GridLayout(2, 1, 0, 0));
			tablePanel.add(tableFavouriteDoublesOther);
		}
		
		tablePanel.repaint();
	}
	
	@Override
	public void valueChanged(ListSelectionEvent arg0)
	{
		JFreeChart pieChart = pieChartPanel.getChart();
		PiePlot plot = (PiePlot)pieChart.getPlot();
		if (selectedScore != null)
		{
			plot.setExplodePercent(selectedScore, 0);
		}
		
		int selectedRow = tableFavouriteDoubles.getSelectedModelRow();
		if (selectedRow == -1)
		{
			return;
		}
		
		selectedScore = (Integer)tableFavouriteDoubles.getValueAt(selectedRow, 0);
		plot.setExplodePercent(selectedScore, 0.2);
	}
}

File: Darts/src/burlton/dartzee/code/screen/stats/player/x01/StatisticsTabFinishBreakdown.kt
package burlton.dartzee.code.screen.stats.player.x01

import burlton.core.code.util.MathsUtil.Companion.getPercentage
import burlton.dartzee.code.screen.stats.player.AbstractStatisticsTab
import burlton.dartzee.code.stats.GameWrapper
import burlton.dartzee.code.utils.getCheckoutSingles
import burlton.desktopcore.code.bean.RowSelectionListener
import burlton.desktopcore.code.bean.ScrollTable
import burlton.desktopcore.code.util.TableUtil.DefaultModel
import burlton.desktopcore.code.util.containsComponent
import org.jfree.chart.ChartFactory
import org.jfree.chart.ChartPanel
import org.jfree.chart.plot.PiePlot
import org.jfree.data.general.DefaultPieDataset
import java.awt.Color
import java.awt.GridLayout
import javax.swing.JPanel
import javax.swing.ListSelectionModel
import javax.swing.border.TitledBorder


class StatisticsTabFinishBreakdown: AbstractStatisticsTab(), RowSelectionListener
{
    private var selectedScore: Int? = null

    private val tableFavouriteDoubles = ScrollTable()
    private val tableFavouriteDoublesOther = ScrollTable()
    private val tablePanel = JPanel()
    private val pieChartPanel = ChartPanel(null)

    init
	{
		layout = GridLayout(1, 3, 0, 0)

		tableFavouriteDoublesOther.tableForeground = Color.RED
		tableFavouriteDoubles.setSelectionMode(ListSelectionModel.SINGLE_SELECTION)
		tableFavouriteDoublesOther.setSelectionMode(ListSelectionModel.SINGLE_SELECTION)
		tablePanel.border = TitledBorder(null, "Double Finishes")

		add(tablePanel)
		tablePanel.layout = GridLayout(2, 1, 0, 0)
		tablePanel.add(tableFavouriteDoubles)
		tablePanel.add(tableFavouriteDoublesOther)
		add(pieChartPanel)

        tableFavouriteDoubles.addRowSelectionListener(this)
	}

    override fun populateStats()
    {
        setTableVisibility()

        val dataset = buildFavouriteDoublesData(tableFavouriteDoubles, filteredGames)
        if (includeOtherComparison())
        {
            buildFavouriteDoublesData(tableFavouriteDoublesOther, filteredGamesOther)
        }

        val pieChart = ChartFactory.createPieChart("Finishes", dataset, true, true, false)
        val plot = pieChart.plot as PiePlot
        plot.labelGenerator = null
        pieChartPanel.chart = pieChart
    }

    private fun buildFavouriteDoublesData(table: ScrollTable, filteredGames: List<GameWrapper>): DefaultPieDataset
    {
        val model = DefaultModel()
        model.addColumn("Double")
        model.addColumn("Finishes")
        model.addColumn("%")
        val dataset = populateFavouriteDoubles(model, filteredGames)
        table.model = model
        table.sortBy(2, true)
        return dataset
    }

    private fun populateFavouriteDoubles(model: DefaultModel, filteredGames: List<GameWrapper>): DefaultPieDataset
    {
        val scores = filteredGames.filter { it.isFinished() }.map { it.getDartsForFinalRound().last().score }

        val rows = scores.distinct().map {
            val count = scores.count { score -> score == it }
            val percent = getPercentage(count, scores.size.toDouble())
            arrayOf(it, count, percent)
        }

        model.addRows(rows)

        //Build up the pie set. Unlike the table, we need ALL values
        val dataset = DefaultPieDataset()
        getCheckoutSingles().sorted().forEach {
            val count = scores.count { score -> score == it }
            dataset.setValue(it.toString(), count)
        }

        return dataset
    }

    private fun setTableVisibility()
    {
        if (!includeOtherComparison())
        {
            tablePanel.layout = GridLayout(1, 1, 0, 0)
            tablePanel.remove(tableFavouriteDoublesOther)
        }
        else if (!containsComponent(tablePanel, tableFavouriteDoublesOther))
        {
            tablePanel.layout = GridLayout(2, 1, 0, 0)
            tablePanel.add(tableFavouriteDoublesOther)
        }

        tablePanel.repaint()
    }

    override fun selectionChanged(src: ScrollTable)
    {
        val pieChart = pieChartPanel.chart
        val plot = pieChart.plot as PiePlot

        //Unset the old value
        selectedScore?.let { plot.setExplodePercent(it.toString(), 0.0) }

        val selectedRow = src.selectedModelRow
        if (selectedRow > -1)
        {
            val newSelection = src.getValueAt(selectedRow, 0) as Int
            plot.setExplodePercent(newSelection.toString(), 0.2)
            selectedScore = newSelection
        }
    }
}